common/cuda_hip/matrix/batch_ell_kernels.hpp.inc

```diff
@@ -0,0 +1,241 @@
+    const gko::batch_dense::BatchEntry<const ValueType>& bobj,
+    const ValueType beta, const gko::batch_dense::BatchEntry<ValueType>& cobj)
+{
+    const auto num_rows = a.num_rows;
+    const auto num_stored_elements_per_row = a.num_stored_elems_per_row;
+    const auto stride = a.stride;
+    const auto b = bobj.values;
+    const auto val = a.values;
+    const auto col = a.col_idxs;
+    const auto c_stride = cobj.stride;
+    auto c = cobj.values;
+    const auto tidx = threadIdx.x;  // thread::get_thread_id_flat();
+    // Specialize the num_thread_per_worker = 1. It doesn't need the shared
+    // memory, __syncthreads, and atomic_add
+    if (tidx < num_rows) {
+        auto temp = zero<ValueType>();
+        for (size_type idx = 0; idx < num_stored_elements_per_row; idx++) {
+            const auto ind = tidx + idx * stride;
+            const auto col_idx = col[ind];
+            if (col_idx < idx) {
+                break;
+            } else {
+                temp += alpha * val[ind] * b[col_idx * bobj.stride];
+            }
+        }
+        const auto c_ind = tidx * c_stride;
+        c[c_ind] = temp + beta * c[c_ind];
+    }
+}
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(default_block_size, sm_multiplier) void spmv(
+    const gko::batch_ell::UniformBatch<const ValueType> a,
+    const batch_dense::UniformBatch<const ValueType> b,
+    const batch_dense::UniformBatch<ValueType> c)
+{
+    for (size_type ibatch = blockIdx.x; ibatch < a.num_batch;
+         ibatch += gridDim.x) {
+        const auto a_b = gko::batch::batch_entry(a, ibatch);
+        const auto b_b = gko::batch::batch_entry(b, ibatch);
+        const auto c_b = gko::batch::batch_entry(c, ibatch);
+        spmv_kernel(a_b, b_b, c_b);
+    }
+}
+
+template <typename ValueType>
+__global__
+    __launch_bounds__(default_block_size, sm_multiplier) void advanced_spmv(
+        const batch_dense::UniformBatch<const ValueType> alpha,
+        const gko::batch_ell::UniformBatch<const ValueType> a,
+        const batch_dense::UniformBatch<const ValueType> b,
+        const batch_dense::UniformBatch<const ValueType> beta,
+        const batch_dense::UniformBatch<ValueType> c)
+{
+    for (size_type ibatch = blockIdx.x; ibatch < a.num_batch;
+         ibatch += gridDim.x) {
+        const auto a_b = gko::batch::batch_entry(a, ibatch);
+        const auto b_b = gko::batch::batch_entry(b, ibatch);
+        const auto c_b = gko::batch::batch_entry(c, ibatch);
+        const auto alpha_b = gko::batch::batch_entry(alpha, ibatch);
+        const auto beta_b = gko::batch::batch_entry(beta, ibatch);
+        const ValueType alphav = alpha_b.values[0];
+        const ValueType betav = beta_b.values[0];
+        advanced_spmv_kernel(alphav, a_b, b_b, betav, c_b);
+    }
+}
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(config::max_block_size) void initialize_zero_dense(
```

Comment:
I am not sure here. Because the idea was to use these only in BatchEll. At least currently, we don't use these in batch_dense. At some point if we do, then we can unify them.

Reply:
If it's not too annoying you could just move it; I can see this being useful in general.